File: Applied Course/4.Problem Solving/5.Problems on Trees/24.Print Right View of a Binary Tree.cpp
#include<bits/stdc++.h>
using namespace std;
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// Logic-1
class Solution {
public:
    vector<int> v;
    vector<int> rightSideView(TreeNode* root) {
        levelOrder(root,0);
        return v;
    }
    void levelOrder(TreeNode* root,int level)
    {
        if(root==NULL)
            return;
        if(v.size()==level)
            v.push_back(root->val);
        levelOrder(root->right,level+1);
        levelOrder(root->left,level+1);
    }
};
// Logic-2
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> v;
        if(!root)
            return v;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty())
        {
            int count=q.size();
            while(count-->0)
            {
                TreeNode* temp=q.front();
                q.pop();
                if(count==0)
                    v.push_back(temp->val);
                if(temp->left!=NULL)
                    q.push(temp->left);
                if(temp->right!=NULL)
                    q.push(temp->right);
            }
        }
        return v;
    }
};
int main()
{
    // https://leetcode.com/problems/binary-tree-right-side-view
    return 0;
}

File: Applied Course/4.Problem Solving/5.Problems on Trees/24.Print Right View of a Binary Tree.py
# https://leetcode.com/problems/binary-tree-right-side-view/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        if not root:
            return root
        queue=[]
        result=[]
        queue.append(root)
        while queue:
            n=len(queue)
            for i in range(1,n+1):
                current=queue[0]
                queue.pop(0)
                if i==n:
                    result.append(current.val)
                if current.left:
                    queue.append(current.left)
                if current.right:
                    queue.append(current.right)
        return result